### PR TITLE
Remove batch class mask calls - Follow up

### PR DESCRIPTION
**Pull request by carriepl**
New, and hopefully improved, version of #143.

In this version, `GFlowNet.sample_actions()` is modified such that : 
- After actions are taken on the environments, new actions masks are computed
- These masks are given to Batch.add_to_batch to avoid recomputation in the batch class
- These masks are given to Env.sample_actions() the next time actions are sampled on these environments.

I tested on the ctorus environment with the following command : 
`python main.py user=$USER +experiments=icml23/ctorus device=cpu env.reward_beta=1`
and validated that it runs and that it produces the exact same graphs of metrics (loss, ...) over iterations in WandB with and without the PR

I also tested on the Tetris environments with the following command :
`python main.py user=$USER env=tetris proxy=tetris gflownet=trajectorybalance env.width=10 env.height=10 logger.do.online=False`
Without this PR : ~7.2 seconds/iteration
With this PR : ~4.8 seconds/iteration

### CODE REVIEW

File: gflownet/gflownet.py [hunk 1]
```diff
@@ -242,14 +247,22 @@ def sample_actions(
...
+            else:
+                mask_invalid_actions = tbool(
+                    [env.get_mask_invalid_actions_backward() for env in envs],
+                    device=self.device,
+                )
         else:
+            # Invalid action masks are provided, convert to pytorch tensors
             mask_invalid_actions = tbool(
-                [env.get_mask_invalid_actions_backward() for env in envs],
+                [m for m in mask_invalid_actions],
```
Comment (alexhernandezgarcia) [inline gflownet/gflownet.py:265]: Why do you need to iterate over `mask_invalid_actions` instead of converting calling `tbool` directly on it?
```suggestion
                mask_invalid_actions,
```
(the suggestion would need a `black` though.

File: gflownet/gflownet.py [hunk 2]
```diff
@@ -424,6 +438,7 @@ def sample_batch(
                         is_forward=True,
                         temperature=1.0,
                         random_action_prob=self.random_action_prob,
+                        mask_invalid_actions=masks_invalid_actions_forward,
```
Comment (alexhernandezgarcia) [inline gflownet/gflownet.py:441]: I would simply pass `None` and remove the definition of `masks_invalid_actions_forward = None` in line 428 above.

File: gflownet/gflownet.py [hunk 3]
```diff
@@ -434,14 +449,39 @@ def sample_batch(
                         is_forward=True,
                         temperature=self.temperature_logits,
                         random_action_prob=self.random_action_prob,
+                        mask_invalid_actions=masks_invalid_actions_forward,
```
Comment (alexhernandezgarcia) [inline gflownet/gflownet.py:452]: Idem. I would simply pass `None` and remove the definition of `masks_invalid_actions_forward = None` in line 428 above.

**Review by alexhernandezgarcia — APPROVED** [review]
See my (minor) comment, but it looks good to me.

As a note, I am currently making some changes in the Batch that will make it more flexible and among other things enable easily retrieving the masks from it. I think that will simplify the code a bit in the future, but for now this is a good solution!

File: gflownet/gflownet.py (hunk 1 above)
Comment (carriepl-mila) [reply gflownet/gflownet.py:265]: Ha! I was definitely half asleep here. You're absolutely right, I'll change that.

Comment (carriepl-mila) [reply gflownet/gflownet.py]: Done.

**Comment by carriepl-mila** [comment]
> See my (minor) comment, but it looks good to me.
> 
> As a note, I am currently making some changes in the Batch that will make it more flexible and among other things enable easily retrieving the masks from it. I think that will simplify the code a bit in the future, but for now this is a good solution!

I think I could have also modified step() to take the masks as optional inputs but I didn't do it because it felt redundant with your PR that was merged recently.  I guess the advantage would be that it would be a little safer than completely skipping the mask checking.
